ecl/hqlcpp/hqlsource.cpp

```diff
@@ -4019,8 +4019,39 @@ void MonitorExtractor::buildKeySegmentInExpr(BuildMonitorState & buildState, Key
 
             HqlExprArray args;
             args.append(*LINK(targetVar));
-            args.append(*LINK(address));
-            args.append(*LINK(address));
+            unsigned srcSize = address->queryType()->getSize();
+            if (srcSize < curSize)
+            {
+                OwnedHqlExpr rangeLower = subctx.getTempDeclare(LINK(fieldType), NULL);
```

Comment:
Look at MonitorExtractor::getRangeLimit.  You should be able to use the code it contains.